R/mlaps.R

```diff
@@ -0,0 +1,394 @@
+    lower_points <- points[1]
+    upper_points <- points[length(points)]
+    pts[is.na(pts) & startsWith(x, "<") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) <= lower_limit] <- lower_points
+    pts[is.na(pts) & startsWith(x, ">") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) >= upper_limit] <- upper_points
+  }
+
+  return(as.numeric(pts))
+}
+
+
+#' @title
+#' Loop mLAPS
+#'
+#' @description
+#' A wrapper around the `mlaps()` function which breaks down the calculation of mLAPS on a
+#' large number of encounters by hospital-year. This avoids memory issues that can be caused
+#' by loading large chunks of the lab table.
+#'
+#' @param db (`DBIConnection`)\cr
+#' RPostgres DB connection.
+#'
+#' @param cohort (`numeric`)\cr
+#' A numeric vector of `genc_ids` to restrict the calculation to.
+#'
+#' @param hours_after_admission (`numeric`)\cr
+#' Consider lab tests collected **up to** `hours_after_admission` hours after inpatient admission in the calculation.
+#' Default `hours_after_admission` is set to 0, where only lab tests collected at Emergency Department (before inpatient admission) is considered in mLAPS calculation.
+#' Since not all encounters are admitted through Emergency Department, depending on research question, it can be relevant to consider lab tests collected in early inpatient admission.
+#' Typically, `hours_after_admission` can be set to 24 to consider any lab tests collected at Emergency Department and 24 hours after inpatient admission.
+#'
+#' @param output_laps_components (`logical`)\cr
+#' Does not aggregate the score and instead outputs for each LAPS component (test), its contribution
+#' to the overall score.
+#'
+#' @return (`data.frame`)\cr
+#' If `output_laps_components == TRUE`:
+#'     `genc_id` (`numeric`),\cr
+#'     `test_type_mapped_omop` (`character`),\cr
+#'     `mlaps` (`numeric`) max score for this test.
+#'
+#' If `output_laps_components == FALSE`:
+#'     `genc_id` (`numeric`),\cr
+#'     `mlaps` (`numeric`) sum of max scores for each relevant test for this encounter.
+#'
+#' @import DBI RPostgreSQL
+#' @export
+#'
+#' @references
+#' When the function is used, please cite the following:
+#' https://doi.org/10.1097/MLR.0b013e3181589bb6
+#' https://doi.org/10.1007/s11606-023-08245-w
+#' https://doi.org/10.1101/2023.01.06.23284273
+#'
+loop_mlaps <- function(db, cohort = NULL, hours_after_admission = 0, output_laps_components = FALSE) {
+
+  hospital_field <- return_hospital_field(db)
+
+  admdad <- DBI::dbGetQuery(
+    db,
+    paste(
+      "SELECT
+        genc_id,
+        admission_date_time,
+        EXTRACT(YEAR FROM discharge_date_time::DATE) AS year,",
+        hospital_field, "AS hospital_id",
+      "FROM admdad",
+      if (!is.null(cohort)) {
+        paste("WHERE genc_id IN (", paste(cohort, collapse = ", "), ")")
+      }
+    )
+  ) %>%
+    as.data.table()
+
+  hospital_years <- admdad %>%
+    select(hospital_id, year) %>%
+    unique()
+
+  res <- purrr::map2_df(
```

Comment:
The `purrr` import is missing at the package level. Needs to be added to the `Imports` in the `DESCRIPTION` file.